src/encoded/static/components/viz/Workflow/Graph.js

```diff
@@ -150,126 +115,185 @@ export default class Graph extends React.Component {
             .reduce(function(maxCount, nodeSet){
                 return Math.max(nodeSet[1].length, maxCount);
             }, 0)
-            .value() * (this.props.rowSpacing) - this.props.rowSpacing,
-            this.props.minimumHeight
+            .value() * (rowSpacing) - rowSpacing,
+            minimumHeight
         );
-    }
+    });
 
-    scrollableWidth(){
-        return (_.reduce(this.props.nodes, function(highestCol, node){
+    static getScrollableWidthFromNodes = memoize(function(nodes, columnWidth, columnSpacing, innerMargin){
+        return (_.reduce(nodes, function(highestCol, node){
             return Math.max(node.column, highestCol);
-        }, 0) + 1) * (this.props.columnWidth + this.props.columnSpacing) + this.props.innerMargin.left + this.props.innerMargin.right - this.props.columnSpacing;
-    }
-
-    nodesWithCoordinates(nodes = null, viewportWidth = null, contentWidth = null, contentHeight = null, verticalMargin = 0){
-
-        if (!contentHeight) contentHeight = this.height();
-
-        if (!nodes) nodes = this.props.nodes.slice(0);
+        }, 0) + 1) * (columnWidth + columnSpacing) + (innerMargin.left || 0) + (innerMargin.right || 0) - columnSpacing;
+    });
+
+    /**
+     * Extends each node with X & Y coordinates.
+     *
+     * Converts column placement and position within columns,
+     * along with other chart dimension settings, into X & Y coordinates.
+     *
+     * IMPORTANT:
+     * Returns a new array but _modifies array items in place_.
+     * If need fresh nodes, deep-clone before supplying `props.nodes`.
```

Comment:
Will remove line "If need fresh nodes, deep-clone before supplying `props.nodes`."
This is actually wrong --- we cannot deep-clone anything because then object references will be lost.
The input `nodes` and `edges` props must be in same shape / instance as returned from `parseAnalysisSteps`. They **cannot** be deep-cloned in between.